### ISSUE TITLE
Incorrect Potential Energy with MonteCarloFlexibleBarostat

### ISSUE THREAD

**Issue by wutobias** [openmm/openmm#4548]
When running simulations with `MonteCarloFlexibleBarostat`, I am seeing high potential energies stored in the context after some of the barostat moves. I think that these moves are generally rejected by the barostat but the potential energies are not re-set correctly to their value before the barostat move attempt. The box vectors and positions seem to be re-set correctly, though.

These are the delta energies from the potential energy stored in the context directly after a barostat move and then collected again after setting the positions and boxvectors to their values stored in the context. One can see the spikes in the potential energy trace.

![MonteCarloFlexibleBarostat](https://github.com/openmm/openmm/assets/33414534/514e5615-32bc-44da-94a3-453cf1a4c1aa)

With the isotropic barostat and without any barostat, there is also a bit of a difference in potential energy. My best guess is that  this is expected due to issues of numerical precision?

![MonteCarloBarostat](https://github.com/openmm/openmm/assets/33414534/0fdadfbd-1a7e-4c1a-a212-99442edac8d8)
![NoBarostat](https://github.com/openmm/openmm/assets/33414534/267c83f2-d969-4775-82c6-9669cae41d2b)

Below is the code and files to reproduce the issue. I ran the code with openmm version 8.1.1. The test system is an equilibrated (1 bar, 300 K) box of 1000 Argon atoms.

```
import openmm
from openmm import app, unit

with open("system.xml", "r") as fopen:
    system_xml = fopen.read()
with open("state.xml", "r") as fopen:
    state_xml = fopen.read()

system = openmm.XmlSerializer.deserialize(system_xml)
state  = openmm.XmlSerializer.deserialize(state_xml)
system.addForce(
    openmm.MonteCarloFlexibleBarostat(
        1*unit.atmosphere, 300*unit.kelvin, 25))
    #openmm.MonteCarloBarostat(
    #    1*unit.atmosphere, 300*unit.kelvin, 25))
integrator = openmm.LangevinMiddleIntegrator(
    300*unit.kelvin,
    1/unit.picosecond,
    1*unit.femtosecond)
platform = openmm.Platform.getPlatformByName("CUDA")
simulation = app.Simulation(
    topology=topology,
    system=system,
    platform=platform,
    integrator=integrator)
simulation.context.setPositions(
    state.getPositions())
simulation.context.setPeriodicBoxVectors(
    *state.getPeriodicBoxVectors())
simulation.context.setVelocities(
    state.getVelocities())

density_list = list()
energy_list  = list()
delta_energy_list = list()
mass = 0.*unit.dalton
for atom in topology.atoms():
    mass += atom.element.mass
for _ in range(10000):
    simulation.step(25)
    state   = simulation.context.getState(
        getEnergy=True, getPositions=True, getVelocities=True)
    vol1    = state.getPeriodicBoxVolume()
    potene1 = state.getPotentialEnergy().value_in_unit(unit.kilojoule_per_mole)
    density = mass/vol1/unit.AVOGADRO_CONSTANT_NA
    density = density.value_in_unit(unit.gram/unit.liter)
    density_list.append(density)
    energy_list.append(potene1)

    simulation.context.setPositions(state.getPositions())
    simulation.context.setPeriodicBoxVectors(*state.getPeriodicBoxVectors())
    simulation.context.setVelocities(state.getVelocities())
    state = simulation.context.getState(
        getEnergy=True, getPositions=True, getVelocities=True)
    potene2 = state.getPotentialEnergy().value_in_unit(unit.kilojoule_per_mole)
    delta_energy_list.append(potene2 - potene1)
```

[system.xml.zip](https://github.com/openmm/openmm/files/15454102/system.xml.zip)
[state.xml.zip](https://github.com/openmm/openmm/files/15454104/state.xml.zip)

**Comment by peastman** [comment]
Thanks for the extra information.  I've made some progress toward debugging this, but it involves a lot of complicated interactions between different parts of the code.  It's related to the fact that internally, it translates all the molecules into a single periodic box to maintain accuracy when representing them in single precision.  That requires coordinating several pieces of information: the positions within the single box, the number of box vectors by which each particle has been translated, and of course the box vectors.  And those pieces of information get updated at different times when a barostat move is applied, when a barostat move is rejected, when the wrapping is updated (done automatically at regular intervals), when positions are set externally, etc.  Something is somehow getting out of sync, but only with the flexible barostat.

Can you try the test above with the CPU platform?  It doesn't do the internal wrapping in the same way CUDA and OpenCL do, and it isn't subject to the original bug you reported.  If the incorrect density also doesn't happen with CPU, it's likely it has the same cause.

**Comment by peastman** [comment]
Thanks.  It sounds like it's the same issue then.

**Comment by peastman** [comment]
I finally managed to track this down.  The problem turned out to be somewhere completely different than I was expecting.  When compiling kernels, the CUDA and OpenCL platforms define macros for applying periodic boundary conditions.  They first check the default box to see whether it's rectangular or triclinic.

https://github.com/openmm/openmm/blob/68fd0f6753241288622634a83357c01df29c6f3f/platforms/opencl/src/OpenCLContext.cpp#L437-L441

If the box is rectangular, they can use a simple calculation.

https://github.com/openmm/openmm/blob/68fd0f6753241288622634a83357c01df29c6f3f/platforms/opencl/src/OpenCLContext.cpp#L472-L473

But a triclinic box requires a more complicated calculation.

https://github.com/openmm/openmm/blob/68fd0f6753241288622634a83357c01df29c6f3f/platforms/opencl/src/OpenCLContext.cpp#L443-L450

The assumption is that if the initial box is rectangular, it will always stay rectangular.  That's true with all the other barostats, which only change the box lengths but not angles.  But MonteCarloFlexibleBarostat can change the angles as well.  That's what's happening in your case.  Your default periodic box is rectangular, but then the barostat changes it to become triclinic.

A very simple fix would be for it to check for the presence of a MonteCarloFlexibleBarostat and assume a triclinic box if it's present.  That would address the issue in this case.

But there are other ways the box shape can change.  It's always possible the user could just directly call `setPeriodicBoxVectors()` on the Context to change it from rectangular to triclinic.  That's probably a rare situation, but it still should be handled correctly.  In that case, we really need to recompile all the kernels.  And I really mean all of them.  These macros are defined at the context level, so they can potentially appear in absolutely any kernel.  The only way to make sure they all get recompiled is to reinitialize the context.

We could make `setPeriodicBoxVectors()` do that automatically, but that risks other types of errors.  It's always possible the System has been modified since the Context was created.  Reinitializing it would pull in those modifications, and that should never be done automatically or silently.  So here is what I suggest.

1. If the System contains a MonteCarloFlexibleBarostat, it will assume a triclinic box.
2. `setPeriodicBoxVectors()` will check the shape of the box being set.  If it's triclinic but the Context is assuming a rectangular box, it will throw an exception with an error message something like this.  "The box shape has changed from rectangular to triclinic.  To do this, you must call setDefaultPeriodicBoxVectors() on the System to specify a triclinic default box, then reinitialize the Context."

Does that sound reasonable?

**Comment by wutobias** [comment]
Thanks for looking into this and the detailed summary.

I see why the issue that you described impact the OpenCL and CUDA platforms. What I don't understand is why the MonteCarloFlexibleBarostat does not get the same mean density as the other two barostats on the CPU platform, since the CPU platform should check the box type every step (right?):
https://github.com/openmm/openmm/blob/68fd0f6753241288622634a83357c01df29c6f3f/platforms/cpu/src/CpuKernels.cpp#L188

> We could make setPeriodicBoxVectors() do that automatically, but that risks other types of errors. It's always possible the System has been modified since the Context was created. Reinitializing it would pull in those modifications, and that should never be done automatically or silently. So here is what I suggest.

Both of your suggestions sound reasonable to me.

**Comment by peastman** [comment]
> What I don't understand is why the MonteCarloFlexibleBarostat does not get the same mean density as the other two barostats on the CPU platform

Are you sure it's a real difference and not just noise?  The difference between 1005 and 1007 is tiny, only about 0.2%.  It's very hard to converge any statistical property to that level of precision.  It usually takes a very long simulation.

Try dividing each simulation in half and computing the density from the first half versus the second half.  How much do they differ?

**Comment by wutobias** [comment]
I just re-ran the waterbox simulations (see above) on the CUDA platform by first setting the system's default box vectors to something triclinic to enforce the triclinic distance macros. Then, once the context was created, I set the periodic box vectors to the cubic ones. **I am now getting the same densities for all three barostats.** 

The numbers below are for the CUDA platform

| System      | Barostat    | $<\rho>$ | OpenMM Version |
| ----------- | ----------- | ---------- | -------------- |
| Waterbox    | Istropic    | 1007       | 8.1.1          |
| Waterbox    | Anisotropic | 1007       | 8.1.1          |
| Waterbox    | Flexible    | 1007   | 8.1.1          |

> Are you sure it's a real difference and not just noise? The difference between 1005 and 1007 is tiny, only about 0.2%. It's very hard to converge any statistical property to that level of precision. It usually takes a very long simulation.

You're correct. For the previous simulations with the CPU platform, I was not able to converge to the same average densities  for all three barostats since I used too short simulations. The volume seems to relax much slower with the flexible barostat compared to the other two barostats.

Thanks again for solving this issue.

**Comment by peastman** [comment]
Thanks!  The fix is in #4618.
